text/0033-logging.md

```diff
@@ -0,0 +1,203 @@
+
+This is an important thing to have when systems generate a lot of logs or are under heavy load. It's also a good practice to populate a monitoring dashboard at certain times and, through that, provide an easy way to analyze and understand what happened in the system in a human-readable way.
+
+Internally, Elasticsearch uses `log4j2` to track events and configure the logging policy. Logging levels can be configured on a per-package basis, giving the precision needed to monitor a specific feature or time in a search engine's lifecycle.
+
+A log level can be: `debug`, `info`, `warn`, `error`, `fatal`, or `unknown`.
+
+#### Elastic Entreprise Search
+
+Elasticsearch's SaaS platform, like Algolia, offers a fairly similar endpoint for browsing system-generated logs.
+
+However, it is possible to disable log storage, change the retention period and also filter the logs with more details.
+
+![image](https://www.elastic.co/guide/en/app-search/current/images/guides/log-settings-controls.png)
+*Elastic App Search's Log Retention settings view.*
+
+> It is also possible to do that from a Log Settings API endpoint.
+
+> API Logs track requests and responses at the engine level, while Analytics Logs track requests, clicks and counts.
+
+### IV. Explanation
+
+#### Current Logging behaviour of MeiliSearch (0.20)
+
+MeiliSearch uses `env_logger` to allow to define the log output level from the `RUST_LOG` environment variable. `env_logger` allows to define a specific log level per module if necessary. [See more here](https://docs.rs/env_logger/0.8.3/env_logger/).
+
+#### Logging behaviour for Milli (0.21)
+
+We have decided to keep the use of `env_logger` for Milli/Transplant. However, we will make some changes to make the logging more consistent and versatile.
+
+We will expose a `MEILI_LOG_LEVEL` environment variable that corresponds to `RUST_LOG` in the binary to improve its meaning and clarity to users. A CLI option `--log-level` will also be added to follow our configuration convention.
+
+##### Log Levels
+
+| Level | Description                                                                                                       |
+|-------|-------------------------------------------------------------------------------------------------------------------|
+| ERROR | Everything related to an error. A non blocking error has occured.                                                 |
+| WARN  | Used to warn about an exceptional non-blocking event occurring.                                                   |
+| INFO  | Default Log Level. It displays high level informations of events occuring in the search engine.                   |
+| DEBUG | Used for debugging and development purposes.  More verbose than INFO.                                             |
+| TRACE | Display everything happening at engine level. Not used at the moment.                                             |
```

Comment:
Did we manage to find a solution to implement the TRACE level? As what I remember we could not have this level

Reply:
Hello @bidoubiwa. 

Sorry to answer lately! The trace level is used internally to debug complex issues. Although we can now use it internally, we will not document it to our users at first.

Reply:
Any reason why it will not be documented for users?

Reply:
Yes, debug will be documented! People developing features with us in rust will be informed about this log level by us if they need an elaborate trace to tackle complex issues. We didn't want to document it to not overwhelm the users who won't really be concerned by it directly. The level trace is specifically for rust developers who will contribute to the engine and it will not be mandatory in the vast majority of cases.

Reply:
If the user is overwhelmed he could go back to debug level. I'm not sure not documenting it is that impactful. We also  not documented served feature which is not a good thing 😕. Also if it is documented we can warn user on the overwhelming aspect of trace and we recommend using debug and only switching if debug is not good enough.

Reply:
Thanks @bidoubiwa! You are right about that documentation point. It's better to document it and let people see that when they possibly need it. We can document the `trace` level and providing some explanation that it is useful for rust developers to deal with complex problems that need to dive at the machine level. E.g. for expert only. I will update the specification.

Reply:
More often than not, trace and debug hide issue. Debug and trace assume a knowledge of the codebase, and will be more confusing than helping for the user.

Reply:
Let's make the choice to have exhaustive documentation on log levels and see how it impacts our users negatively or positively. We can expect to have a really small impact on this decision.